### DISCUSSION TITLE
Solution: VSCode and Brownie solidity compiler solc version are not the same.

### DISCUSSION THREAD

**Discussion by n4n0b1t3** [smartcontractkit/full-blockchain-solidity-course-py#1133] [General]
Let's say you have something like this in your code
```
pragma solidity >=0.6.0 <0.9.0;
import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
```

... and you are trying `brownie compile`. You will get the following error.
```
(py39brownie) fund_me>brownie compile
Muster gefunden werden.
Brownie v1.18.1 - Python development framework for Ethereum

New compatible solc version available: 0.6.12
Compiling contracts...
  Solc version: 0.6.12     <======= see here
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
CompilerError: solc returned the following errors:

C:/Users/Zarathustra/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@0.4.0/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol:2:1: ParserError: Source file requires different compiler version (current compiler is 0.6.12+commit.27d51765.Windows.msvc) - note that nightly builds are considered to be strictly less than the released version
pragma solidity ^0.8.0;
^---------------------^
```
**This won't help ...**
Changing the compiler version in VSCode with the solidity plugin via right click
 "Solidity: change workspace compiler version (Remote)" to 0.8.0 or anything above won't help.

Brownie will ignore this and try to get the highest version in the range of your pragma solidity starting point.
`pragma solidity >=0.6.0 <0.9.0;`
The starting point here is 0.6.0 so Brownie will use 0.6.12 and try to compile this file regardless of the settings of VSCode.

**Solution:** 
Edit: When you need to stay in your main version, e.g. so something between 0.6.6 and 0.6.12 because higher versions are not backward compatible, you have to use the carrot character like so:
`pragma solidity ^0.6.6; `
Now the highest solc version in this case 0.6.12 of this main version 0.6.* will be used.

**Special case**
In the code below, you want to use a library that is written in solidity version 0.8.* (see the import statement). 
```
pragma solidity >=0.8.0 <0.9.0;
import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
```

In this case you might need to adapt your code to fit the higher version. Here a typical upwards compatibility bug.
on 0.8.0 a transfer looks like this `payable(msg.sender).transfer(address(this).balance);`
on 0.6.0 a transfer looks like this `msg.sender.transfer(address(this).balance);`

**Comment by cromewar** [comment]
Yep, the compiler of the extension is just something we add to vscode to have linting highlights and syntax correction, however brownie downloads the compiler for the version of the contract separately.
